src/Coordination/Changelog.cpp

```diff
@@ -585,131 +657,844 @@ class ChangelogReader
+    , commit_logs_cache(log_settings.commit_logs_cache_size_threshold)
+    , prefetch_queue(std::numeric_limits<uint64_t>::max())
     , keeper_context(std::move(keeper_context_))
-    , flush_settings(flush_settings_)
+    , log(getLogger("Changelog"))
+{
+    commit_logs_prefetcher = std::make_unique<ThreadFromGlobalPool>([this] { prefetchCommitLogs(); });
+}
+
+LogEntryStorage::~LogEntryStorage()
+{
+    shutdown();
+}
+
+void LogEntryStorage::prefetchCommitLogs()
 {
-    if (auto latest_log_disk = getLatestLogDisk();
-        log_file_settings.force_sync && dynamic_cast<const DiskLocal *>(latest_log_disk.get()) == nullptr)
+    std::shared_ptr<PrefetchInfo> prefetch_info;
+    while (prefetch_queue.pop(prefetch_info))
     {
-        throw DB::Exception(
-            DB::ErrorCodes::BAD_ARGUMENTS,
-            "force_sync is set to true for logs but disk '{}' cannot satisfy such guarantee because it's not of type DiskLocal.\n"
-            "If you want to use force_sync and same disk for all logs, please set keeper_server.log_storage_disk to a local disk.\n"
-            "If you want to use force_sync and different disk only for old logs, please set 'keeper_server.log_storage_disk' to any "
-            "supported disk and 'keeper_server.latest_log_storage_disk' to a local disk.\n"
-            "Otherwise, disable force_sync",
-            latest_log_disk->getName());
+        if (prefetch_info->cancel)
+        {
+            prefetch_info->done = true;
+            prefetch_info->done.notify_all();
+            continue;
+        }
+
+        auto current_index = prefetch_info->commit_prefetch_index_range.first;
+        try
+        {
+            for (const auto & prefetch_file_info : prefetch_info->file_infos)
+            {
+                const auto & [changelog_description, position, count] = prefetch_file_info;
+                std::lock_guard file_lock(changelog_description->file_mutex);
+                auto file = changelog_description->disk->readFile(changelog_description->path, ReadSettings());
+                file->seek(position, SEEK_SET);
+                LOG_TRACE(log, "Prefetching {} log entries from path {}, from position {}", count, changelog_description->path, position);
+                ProfileEvents::increment(ProfileEvents::KeeperLogsPrefetchedEntries, count);
+
+                for (size_t i = 0; i < count; ++i)
+                {
+                    if (prefetch_info->cancel)
+                        break;
+
+                    auto record = readChangelogRecord(*file, changelog_description->path);
+                    auto entry = logEntryFromRecord(record);
+                    if (current_index != record.header.index)
+                        throw Exception(ErrorCodes::LOGICAL_ERROR, "Invalid index prefetched, expected {}, actual {}", current_index, record.header.index);
+
+                    commit_logs_cache.setPrefetchedEntry(record.header.index, std::move(entry), nullptr);
```

Comment:
Sorry, but it's very confusing. Actually here we are going to say that this particular entry is not prefetched anymore and now it is a normal entry (nobody need to wait on it). 

Maybe rewrite it with `future/promise`?

Reply:
> Maybe rewrite it with future/promise?

The problem is that we would need to send all the `promise` objects to prefetch thread while with the current implementation everything is contained inside the same cache map
I can probably store `std::promise` and `std::future`  in the CacheEntry because I basically did reimplementation of it